language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToConstantCodeAction.java

```diff
@@ -92,28 +100,72 @@ public List<CodeAction> getCodeActions(CodeActionContext context,
+        LSClientCapabilities lsClientCapabilities = context.languageServercontext().get(LSClientCapabilities.class);
+        if (isRange(context.range()) || !lsClientCapabilities.getInitializationOptions().isQuickPickSupported()) {
+            
+            // Selection is a range
+            String value = node.toSourceCode().strip();
+            LineRange replaceRange = node.lineRange();
+            String constDeclStr = String.format("const %s %s = %s;%n", typeSymbol.get().signature(), constName, value);
+            TextEdit constDeclEdit = new TextEdit(new Range(constDeclPosition, constDeclPosition), constDeclStr);
+            TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
+                    PositionUtil.toPosition(replaceRange.endLine())), constName);
+
+            return Collections.singletonList(CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_CONSTANT,
+                    List.of(constDeclEdit, replaceEdit), context.fileUri(), CodeActionKind.RefactorExtract));
+        }
+        
+        // Selection is a position
+        List<Node> nodeList = new ArrayList<>();
+        // Identify the sub-expressions to be extracted
+        while (!(node instanceof StatementNode) && !(node instanceof ModuleMemberDeclarationNode)
+                && !(node instanceof ObjectFieldNode) && !nodeValidator.getInvalidNode()) {
+            nodeList.add(node);
+            node = node.parent();
+            node.accept(nodeValidator);
+        }
+        
+        LinkedHashMap<String, List<TextEdit>> textEditMap = getSubExpressionTextEdits(constName, typeSymbol, 
+                constDeclPosition, nodeList);
+        return Collections.singletonList(CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_CONSTANT,
+                    new Command(NAME, EXTRACT_COMMAND, List.of(NAME, context.filePath().toString(), 
+                            textEditMap)), CodeActionKind.RefactorExtract));
     }
 
     @Override
     public String getName() {
         return NAME;
     }
 
+    private LinkedHashMap<String, List<TextEdit>> getSubExpressionTextEdits(String constName, Optional<TypeSymbol> 
+            typeSymbol, Position constDeclPosition, List<Node> nodeList) {
+        
+        LinkedHashMap<String, List<TextEdit>> textEditMap = new LinkedHashMap<>();
+        nodeList.forEach(extractableNode -> {
+            String value = extractableNode.toSourceCode().strip();
+            LineRange replaceRange = extractableNode.lineRange();
+            String constDeclStr = String.format("const %s %s = %s;%n", typeSymbol.get().signature(), constName, value);
+            TextEdit constDeclEdit =
+                    new TextEdit(new Range(constDeclPosition, constDeclPosition), constDeclStr);
+            TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
+                    PositionUtil.toPosition(replaceRange.endLine())), constName);
```

Comment:
```suggestion

private List<TextEdit> getTextEdits(NonTerminalNode node, TypeSymbol typeSymbol, String constDeclStr {
             String value = node.toSourceCode().strip();
            LineRange replaceRange = node.lineRange();
            String constDeclStr = String.format("const %s %s = %s;%n", typeSymbol.get().signature(), constName, value);
            TextEdit constDeclEdit = new TextEdit(new Range(constDeclPosition, constDeclPosition), constDeclStr);
            TextEdit replaceEdit = new TextEdit(new Range(PositionUtil.toPosition(replaceRange.startLine()),
                    PositionUtil.toPosition(replaceRange.endLine())), constName);
                    return List.of(constDeclEdit, replaceEdit);
                    }
```

We can extract this code snippet to something like this. Where we can reuse it between 114-119 lines.